rules-java-ee/addon/src/main/java/org/jboss/windup/rules/apps/javaee/rules/datasource/DiscoverDataSourceAnnotationRuleProvider.java

```diff
@@ -0,0 +1,129 @@
+ *
+ * @author <a href="mailto:dklingenberg@gmail.com">David Klingenberg</a>
+ *
+ */
+@RuleMetadata(phase = InitialAnalysisPhase.class, after = AnalyzeJavaFilesRuleProvider.class)
+public class DiscoverDataSourceAnnotationRuleProvider extends AbstractRuleProvider
+{
+
+    @Override
+    public Configuration getConfiguration(RuleLoaderContext context)
+    {
+        String ruleIDPrefix = getClass().getSimpleName();
+
+        return ConfigurationBuilder.begin()
+                .addRule()
+                .when(JavaClass.references("javax.annotation.sql.DataSourceDefinition").at(TypeReferenceLocation.ANNOTATION))
+                .perform(new AbstractIterationOperation<JavaTypeReferenceModel>()
+                {
+                    @Override
+                    public void perform(GraphRewrite event, EvaluationContext context, JavaTypeReferenceModel payload)
+                    {
+                        extractDataSourceMetadata(event, payload);
+                    }
+                })
+                .withId(ruleIDPrefix + "_DataSourceDefinition");
+    }
+
+    private String getAnnotationLiteralValue(JavaAnnotationTypeReferenceModel model, String name)
+    {
+        JavaAnnotationTypeValueModel valueModel = model.getAnnotationValues().get(name);
+
+        if (valueModel instanceof JavaAnnotationLiteralTypeValueModel)
+        {
+            JavaAnnotationLiteralTypeValueModel literalTypeValue = (JavaAnnotationLiteralTypeValueModel) valueModel;
+            return literalTypeValue.getLiteralValue();
+        }
+        else
+        {
+            return null;
+        }
+    }
+
+    private void extractDataSourceMetadata(GraphRewrite event, JavaTypeReferenceModel javaTypeReference)
+    {
+        javaTypeReference.getFile().setGenerateSourceReport(true);
+        JavaAnnotationTypeReferenceModel annotationTypeReference = (JavaAnnotationTypeReferenceModel) javaTypeReference;
+
+        JavaClassModel datasourceClass = getJavaClass(javaTypeReference);
+
+        String dataSourceName = getAnnotationLiteralValue(annotationTypeReference, "name");
+        if (Strings.isNullOrEmpty(dataSourceName))
+        {
+            dataSourceName = datasourceClass.getClassName();
+        }
+
+        String isXaString = getAnnotationLiteralValue(annotationTypeReference, "transactional");
+
+        boolean isXa = isXaString == null || Boolean.getBoolean(isXaString);
+
+        Service<DataSourceModel> dataSourceService = new GraphService<>(event.getGraphContext(), DataSourceModel.class);
+        DataSourceModel dataSourceModel = dataSourceService.create();
+        Set<ProjectModel> applications = ProjectTraversalCache.getApplicationsForProject(event.getGraphContext(), javaTypeReference.getFile().getProjectModel());
+
+        dataSourceModel.setApplications(applications);
+        dataSourceModel.setName(dataSourceName);
+        dataSourceModel.setXa(isXa);
```

Comment:
Here `dataSourceModel.setJndiLocation(dataSourceName)` should be added because _Server Resource Report_ from [`server.ftl`](https://github.com/windup/windup/blob/master/rules-java-ee/addon/src/main/resources/reports/templates/server.ftl#L58) shows the JNDI location.

Reply:
+1, otherwise it will throw exception on Server resource template